qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,290 @@
+import dataclasses
+from abc import ABC, abstractmethod
+from typing import Dict, Tuple
+
+import gymnasium
+import numpy as np
+
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
+    """Holds statistics about an archive.
+
+    Attributes:
+        coverage: Proportion of cells in the archive that have an elite - always
+            in the range $[0,1]$.
+        qd_score: QD score, i.e. sum of objective values of all elites in the
+            archive. All the fitnesses are normalized to $[0,1]$ using the
+            fitness_bounds before computing the QD score.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+    """
+
+    coverage: float
+    qd_score: float
+    max_fitness: float
+    ccdf: np.ndarray
+
+
+class DiscreteArchiveMetricsCalculator(ABC):
```

Comment:
The first thing I think about when I hear about metrics is a function, not really an object that I have to instantiate. Given that these classes are all essentially executing one function (get_metrics), what do you think of refactoring all of these into functions and helper functions? Or at least, maybe `get_metrics` should become `__call__`.

On the other hand, there are precedents for "calculator" functions that are also classes, e.g. loss functions in PyTorch https://pytorch.org/docs/stable/nn.html#loss-functions -- note that these also use `__call__`.

Reply:
I thought about several structures. The reasons why I chose objects are:
- I like the idea that someone creates an object of type `GridMetricsCalculator`, and gives the parameters (e.g. grid resolution, fitness bounds, feature space) only once, and then use this single object to calculate the metrics for all variants and all replications
- If I used functions and helper functions:
    - the number of arguments could be quite long, for the main function and for the helper functions...
    - if the user wants to fix some parameters (e.g. grid resolution, feature space...) he would have to know and ideally use `partial`. And I don't like `partial`. One of the reasons is that many people do not how it works, and I find it quite ugly. So I try to avoid it where I can.

Reply:
And I don't like using only `__call__` for code that is intended for a wide audience, as many people do not know it.

As an in-between solution, I created a method `__call__` that directly calls `get_metrics` (as done in PyTorch)

Reply:
These are fair points; using `__call__` also opens up the opportunity for us to put wrappers around `get_metrics` like input validation and output validation in the future if the need should arise.